src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -139,385 +111,170 @@ public ImmutableArray<DiagnosticData> GetBuildErrors()
-        var buildErrors = GetBuildErrors().Except(allLiveErrors).GroupBy(k => k.DocumentId);
-        foreach (var group in buildErrors)
-        {
-            cancellationToken.ThrowIfCancellationRequested();
-
-            if (group.Key == null)
-            {
-                foreach (var projectGroup in group.GroupBy(g => g.ProjectId))
-                {
-                    Contract.ThrowIfNull(projectGroup.Key);
-                    argsBuilder.Add(CreateArgsToReportBuildErrors(projectGroup.Key, solution, projectGroup.ToImmutableArray()));
-                }
-
-                continue;
-            }
-
-            argsBuilder.Add(CreateArgsToReportBuildErrors(group.Key, solution, group.ToImmutableArray()));
-        }
-
-        ProcessAndRaiseDiagnosticsUpdated(argsBuilder.ToImmutableAndClear());
-
-        // Report pending live errors
-        return diagnosticService.SynchronizeWithBuildAsync(_workspace, pendingLiveErrorsToSync, onBuildCompleted: true, cancellationToken);
+        _ = ClearInProgressState();
     }
 
-    private static DiagnosticsUpdatedArgs CreateArgsToReportBuildErrors<T>(T item, Solution solution, ImmutableArray<DiagnosticData> buildErrors)
+    public void AddNewErrors(ProjectId projectId, Guid projectHierarchyGuid, ImmutableArray<DiagnosticData> diagnostics)
     {
-        if (item is ProjectId projectId)
-        {
-            return CreateDiagnosticsCreatedArgs(projectId, solution, projectId, documentId: null, buildErrors);
-        }
+        Debug.Assert(diagnostics.All(d => d.IsBuildDiagnostic()));
 
-        RoslynDebug.Assert(item is DocumentId);
-        var documentId = (DocumentId)(object)item;
-        return CreateDiagnosticsCreatedArgs(documentId, solution, documentId.ProjectId, documentId, buildErrors);
+        // Capture state that will be processed in background thread.
+        var state = GetOrCreateInProgressState();
+
+        _taskQueue.AddWork(async cancellationToken =>
+        {
+            await ProcessDiagnosticsReportAsync(projectId, projectHierarchyGuid, diagnostics, state, cancellationToken).ConfigureAwait(false);
+        });
     }
 
-    private static void AddArgsToClearBuildOnlyProjectErrors(ref TemporaryArray<DiagnosticsUpdatedArgs> builder, Solution solution, ProjectId? projectId)
+    private async Task ClearPreviousAsync(ProjectId? projectId, CancellationToken cancellationToken)
     {
-        // Remove all project errors
-        builder.Add(CreateDiagnosticsRemovedArgs(projectId, solution, projectId, documentId: null));
+        var diagnosticManagerService = await GetOrCreateDiagnosticManagerAsync(cancellationToken).ConfigureAwait(false);
 
-        var project = solution.GetProject(projectId);
-        if (project == null)
+        if (projectId is not null)
         {
-            return;
+            await diagnosticManagerService.ClearDiagnosticsAsync(projectId.Id.ToString(), cancellationToken).ConfigureAwait(false);
```

Comment:
doc that the project's guid is how the diag service identifies things?

Reply:
will take in a followup